consensus/hotstuff/cruisectl/block_rate_controller.go

```diff
@@ -0,0 +1,160 @@
+	integralErr     float64   // E_N[v]  - integral of error at view v
+	derivativeErr   float64   // ∆_N[v]  - derivative of error at view v
+}
+
+// epochInfo stores data about the current and next epoch. It is updated when we enter
+// the first view of a new epoch, or the EpochSetup phase of the current epoch.
+type epochInfo struct {
+	curEpochFinalView        uint64
+	curEpochTargetSwitchover time.Time
+	nextEpochFinalView       *uint64
+	epochFallbackTriggered   *atomic.Bool
+}
+
+// BlockRateController dynamically adjusts the block rate delay of this node,
+// based on the measured block rate of the consensus committee as a whole, in
+// order to achieve a target overall block rate.
+type BlockRateController struct {
+	component.Component
+
+	config *Config
+	state  protocol.State
+	log    zerolog.Logger
+
+	lastMeasurement *measurement    // the most recently taken measurement
+	blockRateDelay  *atomic.Float64 // the block rate delay value to use when proposing a block
+	epochInfo
+
+	viewChanges chan uint64       // OnViewChange events           (view entered)
+	epochSetups chan *flow.Header // EpochSetupPhaseStarted events (block header within setup phase)
+}
+
+// NewBlockRateController returns a new BlockRateController.
+func NewBlockRateController(log zerolog.Logger, config *Config, state protocol.State) (*BlockRateController, error) {
+
+	ctl := &BlockRateController{
+		config:      config,
+		log:         log,
+		state:       state,
+		viewChanges: make(chan uint64, 1),
+		epochSetups: make(chan *flow.Header, 1),
+	}
+
+	ctl.Component = component.NewComponentManagerBuilder().
+		AddWorker(ctl.processEventsWorker).
+		Build()
+
+	// TODO initialize last measurement
+	// TODO initialize epoch info
+	_ = ctl.lastMeasurement
+	_ = ctl.curEpochTargetSwitchover
+	_ = ctl.curEpochFinalView
+	_ = ctl.nextEpochFinalView
+
+	return ctl, nil
+}
+
+// BlockRateDelay returns the current block rate delay value to use when proposing, in milliseconds.
+// This function reflects the most recently computed output of the PID controller
+func (ctl *BlockRateController) BlockRateDelay() float64 {
+	return ctl.blockRateDelay.Load()
+}
+
+// processEventsWorker is a worker routine which processes events received from other components.
+func (ctl *BlockRateController) processEventsWorker(ctx irrecoverable.SignalerContext, ready component.ReadyFunc) {
+	ready()
+
+	done := ctx.Done()
+
+	for {
+		select {
+		case <-done:
+			return
+		case enteredView := <-ctl.viewChanges:
+			err := ctl.processOnViewChange(enteredView)
+			if err != nil {
+				ctl.log.Err(err).Msgf("fatal error handling OnViewChange event")
+				ctx.Throw(err)
+			}
+		case block := <-ctl.epochSetups:
+			snapshot := ctl.state.AtHeight(block.Height)
+			err := ctl.processEpochSetupPhaseStarted(snapshot)
+			if err != nil {
+				ctl.log.Err(err).Msgf("fatal error handling EpochSetupPhaseStarted event")
+				ctx.Throw(err)
+			}
+		}
```

Comment:
I think `epochSetups` events need to be processed with strict priority. We want to avoid a node catching up only running through `viewChanges` and not getting to the epoch setup event despite it waiting in the queue.

Reply:
https://github.com/onflow/flow-go/pull/4332/commits/4f5027ef844b6600c46cbe726ac3f73a5eb92f47